torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,465 @@
+
+    if (kv == j.end()) {
+      return false;
+    }
+    if (kv.value().type() != expected_type) {
+      return false;
+    }
+    kv.value().get_to(to);
+    return true;
+  }
+
+  static std::unique_ptr<BoundaryMetadata> Build(
+      const nlohmann::basic_json<>& j) {
+    BoundaryMetadata metadata;
+
+    bool is_valid_metadata_json =
+        CopyJsonValue(j, "name", json::value_t::string, metadata.name) &&
+        CopyJsonValue(j, "id", json::value_t::string, metadata.id) &&
+        CopyJsonValue(j, "pos", json::value_t::number_unsigned, metadata.pos) &&
+        CopyJsonValue(j, "is_input", json::value_t::boolean, metadata.is_input);
+
+    if (!is_valid_metadata_json) {
+      return nullptr;
+    }
+
+    if (auto kv = j.find("attr"); kv != j.end() && kv.value().is_object()) {
+      auto& attrs_j = kv.value();
+      for (auto attr_j = attrs_j.begin(); attr_j != attrs_j.end(); ++attr_j) {
+        metadata.attrs.insert({attr_j.key(), attr_j.value()});
+      }
+    }
+    return std::make_unique<BoundaryMetadata>(std::move(metadata));
+  }
+};
+
+class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
+ public:
+  explicit BuildStableHLOCompositePass()
+      : mlir::OperationPass<mlir::ModuleOp>::OperationPass(
+            mlir::TypeID::get<BuildStableHLOCompositePass>()) {}
+
+  ~BuildStableHLOCompositePass() override = default;
+
+  void runOnOperation() override {
+    mlir::ModuleOp module_op = getOperation();
+    llvm::SmallVector<mlir::func::FuncOp> func_ops(
+        module_op.getOps<mlir::func::FuncOp>());
+    for (mlir::func::FuncOp& func_op : func_ops) {
+      llvm::DenseMap<const mlir::Operation*, size_t> op_order_map =
+          BuildOpOrderMap(func_op);
+      for (auto op : func_op.getOps<mlir::stablehlo::CustomCallOp>()) {
+        if (mlir::failed(
+                BuildStableHLOComposite(op.getOperation(), op_order_map))) {
+          op.emitError() << "failed to build composite.";
+          return signalPassFailure();
+        }
+      }
+    }
+  }
+
+  mlir::StringRef getName() const override {
+    return llvm::getTypeName<BuildStableHLOCompositePass>();
+  }
+
+  std::unique_ptr<mlir::Pass> clonePass() const override {
+    return std::make_unique<BuildStableHLOCompositePass>(*this);
+  }
+
+ private:
+  llvm::DenseMap<const mlir::Operation*, size_t> BuildOpOrderMap(
+      mlir::func::FuncOp func_op) const {
+    llvm::DenseMap<const mlir::Operation*, size_t> op_order_map;
+    for (const auto& op : llvm::enumerate(func_op.getOps())) {
+      op_order_map[&op.value()] = op.index();
+    }
+    return op_order_map;
+  }
+
+  mlir::FailureOr<std::unique_ptr<BoundaryMetadata>> GetBoundaryMetadata(
+      mlir::Operation* op) {
+    if (!IsXlaMarkTensorOp(op)) {
+      return mlir::FailureOr(nullptr);
+    }
+    auto backend_config =
+        op->getAttr("backend_config").dyn_cast<mlir::StringAttr>();
+    if (backend_config == nullptr) {
+      return mlir::FailureOr(nullptr);
```

Comment:
Silent pass failures tend to be difficult to track down. Would recommend using `op->emitError()` for each of these failure cases. Usually we implement failure messages at the "leafs" (i.e. this error is propagated up, and if `GetBoundaryMetadata` fails, the callsite also returns failure, but that one can be silent).

```
return op->emitError() << "expected mark tensor operation";
...
return op->emitError() << "expected mark tensor op to have backend_config";
```

Reply:
This is not a failure though. `unique_ptr` here is used as an alternative of `optional`. So the semantic of `FailureOr(nullptr)` is "the operation succeeded, but got no boundary metadata".